uint64 inc
uint8 pirF
uint8 pirB